src/iop/sparse_linear_algebra.rs

```diff
@@ -0,0 +1,256 @@
+    }
+
+    #[test]
+    fn mat_vec_mul_squat() {
+        let rng = &mut thread_rng();
+
+        let num_rounds = 100;
+        let num_rows_dist = Uniform::from(1..10);
+        let num_cols_dist = Uniform::from(10..100);
+        let fill_factor_dist = Uniform::new(0.0, 1.0);
+
+        for _ in 0..num_rounds {
+            let num_rows = num_rows_dist.sample(rng);
+            let num_cols = num_cols_dist.sample(rng);
+            let fill_factor = fill_factor_dist.sample(rng);
+            let dense_matrix =
+                DenseMatrix::<F>::generate_random(num_rows, num_cols, fill_factor, rng);
+            let sparse_matrix = dense_matrix.to_sparse();
+            let vec = generate_random_vector(num_cols, rng);
+
+            let dense_product = dense_matrix.vec_mul(&vec);
+            let sparse_product = mat_vec_mul(&sparse_matrix, &vec);
+            assert_eq!(dense_product, sparse_product);
+        }
+    }
+
+    #[test]
+    fn mat_vec_mul_shorter_vec() {
+        let rng = &mut thread_rng();
+
+        let num_rounds = 100;
+        let matrix_size_dist = Uniform::from(1..100);
+        let fill_factor_dist = Uniform::new(0.0, 1.0);
+
+        for _ in 0..num_rounds {
+            let num_rows = matrix_size_dist.sample(rng);
+            let num_cols = matrix_size_dist.sample(rng);
+            let fill_factor = fill_factor_dist.sample(rng);
+            let vec_size = rng.gen_range(1..=num_cols);
+            let dense_matrix =
+                DenseMatrix::<F>::generate_random(num_rows, num_cols, fill_factor, rng);
+            let sparse_matrix = dense_matrix.to_sparse();
+            let mut vec = generate_random_vector(num_cols, rng);
+            for i in vec[vec_size..].iter_mut() {
+                *i = F::zero();
+            }
+
+            let dense_product = dense_matrix.vec_mul(&vec);
+            let sparse_product = mat_vec_mul(&sparse_matrix, &vec[0..vec_size]);
+            assert_eq!(dense_product, sparse_product);
+        }
+    }
+
+    #[test]
+    fn mat_vec_mul_longer_vec() {
+        let rng = &mut thread_rng();
+
+        let num_rounds = 100;
+        let matrix_size_dist = Uniform::from(1..100);
+        let fill_factor_dist = Uniform::new(0.0, 1.0);
+
+        for _ in 0..num_rounds {
+            let num_rows = matrix_size_dist.sample(rng);
+            let num_cols = matrix_size_dist.sample(rng);
+            let fill_factor = fill_factor_dist.sample(rng);
+            let border = rng.gen_range(1..=num_cols);
+            let mut dense_matrix =
+                DenseMatrix::<F>::generate_random(num_rows, num_cols, fill_factor, rng);
+            for (k, val) in dense_matrix.val.iter_mut().enumerate() {
+                let j = k % dense_matrix.num_cols;
+                if j > num_cols - border {
+                    *val = F::zero();
+                }
+            }
+            let sparse_matrix = dense_matrix.to_sparse();
+            let vec = generate_random_vector(num_cols, rng);
```

Comment:
Isn't the intention of the test to sample a vector which is longer than `num_cols` entries?

Reply:
Indeed. And it was what the test actually did, but choice of variable names and general logic was a bit misleading. Changed the code to more clearly outline the intention of the test.